casper/src/test/scala/coop/rchain/casper/batch2/BlockReceiverEffectsSpec.scala

```diff
@@ -0,0 +1,234 @@
+    case (incomingQueue, _, outStream, bs, br, bds) =>
+      for {
+        block <- addBlock[Task](bs)
+        _     <- incomingQueue.enqueue1(block)
+      } yield {
+        bs.put(Seq((block.blockHash, block))) wasCalled once
+        bs.contains(*) wasNever called
+        br.ackReceived(*) wasNever called
+        dagStorageWasNotModified(bds)
+        outStream should notEmit
+      }
+  }
+
+  it should "pass to output blocks with resolved dependencies" in withEnv[Task]("root") {
+    case (incomingQueue, validatedQueue, outStream, bs, br, bds) =>
+      for {
+        // Received a parent with an empty list of justifications and its child
+        a1 <- Sync[Task].delay(makeBlock())
+        a2 = makeBlock(List(a1.blockHash))
+
+        // Put the parent and child in the input queue
+        _ <- incomingQueue.enqueue1(a2)
+        _ <- incomingQueue.enqueue1(a1)
+
+        // Dependencies of the child (its parent) have not yet been resolved,
+        // so only the parent goes to the output queue, since it has no dependencies
+        a1InOutQueue <- outStream.take(1).compile.lastOrError
+
+        // A1 is now validated (e.g. in BlockProcessor)
+        _ <- validatedQueue.enqueue1(a1)
+
+        // All dependencies of child A2 are resolved, so it also goes to the output queue
+        a2InOutQueue <- outStream.take(1).compile.lastOrError
+      } yield {
+        val bsPutCaptor = ArgCaptor[Seq[(BlockHash, BlockMessage)]]
+        bs.put(bsPutCaptor) wasCalled twice
+        bsPutCaptor.values should contain allOf (Seq((a1.blockHash, a1)), Seq((a2.blockHash, a2)))
+
+        val bsContainsCaptor = ArgCaptor[Seq[BlockHash]]
+        bs.contains(bsContainsCaptor) wasCalled 3.times
+        bsContainsCaptor.values should contain allOf (Seq(a1.blockHash), Seq(a2.blockHash))
+
+        val brAckReceivedCaptor = ArgCaptor[BlockHash]
+        br.ackReceived(brAckReceivedCaptor) wasCalled twice
+        brAckReceivedCaptor.values should contain allOf (a1.blockHash, a2.blockHash)
+
+        dagStorageWasNotModified(bds)
+        a1InOutQueue shouldBe a1.blockHash
+        a2InOutQueue shouldBe a2.blockHash
+      }
+  }
+
+  private def blockDagStorageMock[F[_]: Applicative]: BlockDagStorage[F] = {
+    val emptyDag = DagRepresentation(Set(), Map(), SortedMap(), DagMessageState(), Map())
+    mock[BlockDagStorage[F]].getRepresentation returnsF emptyDag
+  }
+
+  private def blockRetrieverMock[F[_]: Applicative]: BlockRetriever[F] =
+    mock[BlockRetriever[F]].ackReceived(*) returns ().pure[F]
+
+  private def blockStoreMock[F[_]: Sync]: BlockStore[F] = {
+    val state  = Ref.unsafe[F, Map[BlockHash, BlockMessage]](Map())
+    val bsMock = mock[BlockStore[F]]
+    bsMock.contains(*) answers { keys: Seq[BlockHash] =>
+      state.get.map(s => Seq(s.contains(keys.head)))
+    }
+    bsMock.put(*) answers { kvPairs: Seq[(BlockHash, BlockMessage)] =>
+      state.update(s => kvPairs.foldLeft(s) { case (acc, item) => acc + item })
+    }
```

Comment:
I missed these two cases. You have `F[_]` expression _in the air_, they are not connected to F execution so it's just like a declaration and not execution. All F values must be composed in a sequence to be executed.
This is probably the source for the null reference error you reported.

Reply:
@tgrospic, yes, you are right. I tried to put mocks creation in F-context, but it didn't have effect.